### PR TITLE
Better CLI argument handling & more documentation

### LINKED ISSUE #78: YAFC fails to launch from a command line without an existing project
Use case:
You're a new user and you want to launch YAFC from the command line. You can't do that because without the `project` argument, `args.Length < 3` becomes `true`, which leads to YAFC not loading.

I didn't reproduce this case myself, and it's low priority due to the low number of console users.  
I'm creating this issue to not forget that this bug exists.

### PR DESCRIPTION
**Pull request by shihan42**
I've completely overhauled the handling and interpretation of command line arguments, along with updating the documentation and adding examples to illustrate the different arguments and their respective use cases.

There was a potential error in the code related to handling recent projects, which could result in attempting to load a project that might be null. This has been fixed (hopefully).

Closes #78

### CODE REVIEW

File: Docs/CLI.md
Comment (veger) [inline Docs/CLI.md:5]:
```suggestion
`YAFC [<data-path> [--mods-path <path>] [--project-file <path>] [--expensive]] [--help]`
```

File: YAFC/Program.cs
Comment (veger) [inline YAFC/Program.cs:42]: It would be nice to return with a non-zero exit code when there is an error. (when people make scripts to start YAFC the script can detect failures)

File: YAFC/Utils/CommandLineParser.cs
Comment (veger) [inline YAFC/Utils/CommandLineParser.cs:1]: Does it make sense to use [`System.CommandLine`](https://learn.microsoft.com/en-us/dotnet/standard/commandline/)? It is still in beta/testing, but I would assume it becomes a standard at some point?

File: YAFC/Utils/CommandLineParser.cs
Comment (veger) [inline YAFC/Utils/CommandLineParser.cs:29]: > !args[i + 1].StartsWith("--")

I could name my folder `--factorio-test-mods`? I do not think there is a need for this check as it will work fine without, or check if this folder exists and only return the missing argument error when it does not.

File: YAFC/Utils/CommandLineParser.cs
Comment (veger) [inline YAFC/Utils/CommandLineParser.cs:44]: same here

File: YAFC/Utils/CommandLineParser.cs
Comment (veger) [inline YAFC/Utils/CommandLineParser.cs:77]:
```suggestion
            Console.WriteLine("YAFC [[<data-path>] [--mods-path <path>] [--project-file <path>] [--expensive]] [--help]");
```

File: YAFC/Windows/WelcomeScreen.cs
Comment (veger) [inline YAFC/Windows/WelcomeScreen.cs:75]: `cliProject` is never `null` as you now set it with

> ProjectDefinition cliProject = CommandLineParser.Parse(args);

**Review by veger — COMMENTED** [review]
nice improvements!

File: Docs/CLI.md
Comment (shihan42) [reply Docs/CLI.md:5]: Yeah, that makes it cleared that there are two modes of operation.

File: YAFC/Windows/WelcomeScreen.cs
Comment (shihan42) [reply YAFC/Windows/WelcomeScreen.cs:75]: It can be, since that's the default value for the parameter. If only `Program.cs` would instantiate a new `WelcomeScreen`, we'd be safe. But there is the option to return to the welcome screen from the main window ;-)
https://github.com/have-fun-was-taken/yafc-ce/blob/aa077c816c9ea80dec4dff612bf1855b54a339eb/YAFC/Windows/MainScreen.cs#L652

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs:77]: Done.

File: YAFC/Program.cs
Comment (shihan42) [reply YAFC/Program.cs:42]: Now it returns the ubiquituous **1** :-)

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs:1]: I thought about it, since it seems legit. But in the end, I decided to wait for it to become a standard, production-ready part of .Net.

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs]: That is evil 😄 

But now entirely possible 😉

Comment (veger) [reply YAFC/Utils/CommandLineParser.cs]: It certainly is evil, sorry about that :smiling_imp:

But it was to make a point about limiting the use cases without (a good) reason. And usually, at some point there will be a user impacted and asking about it :smile:

File: YAFC/Windows/WelcomeScreen.cs
Comment (veger) [reply YAFC/Windows/WelcomeScreen.cs:75]: A nice catch, forgot about this path to the welcome screen! :+1:

**Review by veger — DISMISSED** [review]

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs]: Don't you dare 😨 

Here would be the point where I can only tell the user: Use at your own risk, be nice, and don't use silly directory names 🤣

File: YAFC/Utils/CommandLineParser.cs
Comment (have-fun-was-taken) [inline YAFC/Utils/CommandLineParser.cs:19]: I think it would be better to first validate the value and then to assign it because in this case we don't assign incorrect data to the `options`. The PR's implementation has a potential for trouble if the inserted data is automatically propagated somewhere else in the future. That applies to all parameters.
The pseudocode is something like
```
if (!validationCondition) {
report error;
return null;
}
assign option property;
```

File: YAFC/Utils/CommandLineParser.cs
Comment (have-fun-was-taken) [inline YAFC/Utils/CommandLineParser.cs:76]: That multiline print can be improved by using [verbatim](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/#verbatim-string-literals) string literals. It allows to write multiline outputs without `WriteLine` for each line.
```
Console.WriteLine("Usage:
YAFC");

/* Output: 
Usage:
YAFC
*/
```

File: YAFC/Windows/WelcomeScreen.cs
Comment (have-fun-was-taken) [inline YAFC/Windows/WelcomeScreen.cs:283]: When the method names are a verb in its base form -- `SetProject` -- the documentation of the method starts with a verb in its third form -- `Initializes the different [...]`.

File: YAFC/Utils/CommandLineParser.cs
Comment (have-fun-was-taken) [reply YAFC/Utils/CommandLineParser.cs:76]: Update: fixed the comment above -- forgot @.

**Comment by have-fun-was-taken** [comment]
I updated the merge rules -- a PR-approval is dismissed if a commit is pushed after the approval. That will help to us to review the whole code that is considered for a merge.

**Comment by shihan42** [comment]
> I updated the merge rules -- a PR-approval is dismissed if a commit is pushed after the approval. That will help to us to review the whole code that is considered for a merge.

Great idea 👍 

I'll work on your comments today or tomorrow.

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs]: Done.

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs:19]: Right now, the project object is directly used in `WelcomeScreen.cs`, which has a validator with `WelcomeScreen.ValidateSelection()`.
So, validating in the `CommandLineParser` would lead to redundant code. Since `WelcomeScreen` has different ways to populate a `ProjectDefinition`, we can't simply move the validation code into `CommandLineParser`.

But what we can do is to move all the validation code into `ProjectDefinition`, since that would adhere better to Single Responsibility. From there, `CommandLineParser`, `WelcomeScreen`, and any other class could use it without a hassle.

What do you think?

Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs:19]: Addendum:
People keep experiencing crashes and we strongly suspect that it's because of `ProjectDefinition`s that are `null`, don't we?
So I'd like to get this PR on the road to mitigate this.
I'd submit a followup PR that refactors the validation, if needed/desired.

**Comment by shihan42** [comment]
Sorry, RL hit hard this week.
But now, I'm back on track.

**Comment by have-fun-was-taken** [comment]
No problem, I understand.

File: YAFC/Utils/CommandLineParser.cs
Comment (have-fun-was-taken) [reply YAFC/Utils/CommandLineParser.cs:19]: > People keep experiencing crashes

If you're talking about the crashes mentioned [there](https://github.com/have-fun-was-taken/yafc-ce/issues/82), it's because the [fix](https://github.com/have-fun-was-taken/yafc-ce/pull/93) was not in the prevoius release. I've made [RC2](https://github.com/have-fun-was-taken/yafc-ce/releases/tag/0.6.3-RC2) yesterday that includes it.

Comment (have-fun-was-taken) [reply YAFC/Utils/CommandLineParser.cs:19]: > move all the validation code into ProjectDefinition

Good idea! I'll summarize it to see if I got it correctly. 

`CommandLineParser` is responsible for parsing the CLI arguments and returning a `ProjectDefinition` instance.  
The class `ProjectDefinition` checks the correctness of the input data in its own constructor.

**Comment by have-fun-was-taken** [comment]
It would be good to have a small note on these improvements in the changelog.

File: YAFC/Utils/CommandLineParser.cs
Comment (shihan42) [reply YAFC/Utils/CommandLineParser.cs:19]: Exactly. Plus the possibility to create other `ProjectDefinition`s (e.g. in the list of recent projects). Regardless of the source, the objects then can validate themselves.

Documentation is a good hint, I tend to forget it 😉

**Comment by veger** [comment]
What is the status on this PR? Is it ready to be merged or are there other improvements planned?

**Comment by shihan42** [comment]
After reading the conversations again, it seems that I'm waiting for an answer to
> So I'd like to get this PR on the road to mitigate this.
> I'd submit a followup PR that refactors the validation, if needed/desired.
> https://github.com/have-fun-was-taken/yafc-ce/pull/89#discussion_r1562144951

Apart from that, the PR is done, I think.

**Review by have-fun-was-taken — APPROVED** [review]
Force-pushed because I fixed the merge conflict by rebase

**Pull request merged**
